### PR TITLE
[1단계 - 자동차 경주 구현] 에버(손채영) 미션 제출합니다.

### PR DESCRIPTION
**Pull request by helenason**
안녕하세요! 6기 에버입니다.

이번 자동차 경주 미션은 [@리비](https://github.com/Libienz) 와 페어로 진행한 미션입니다.

게임에 대해서는 다음의 추가적인 설정을 저희끼리 임의로 정하였음을 알려드립니다.

- 시도 횟수로 0을 받을 수 없게 설정하였습니다.
- 유저 입력에서 검증 오류 발생 시 재입력을 받도록 설계하였습니다.

또 간단하게 설계 포인트 아래에 남겨드리니 참고하여 리뷰해주시면 더욱 감사하겠습니다 ☺️

### 설계 포인트

- 객체지향 생활체조 7원칙을 기본 바탕으로 원칙을 준수하며 작업하려고 노력했습니다.
- 다만 원칙을 지킴으로써 얻는 코드 퀄리티보다 코드 복잡성이 크다고 판단한 경우는 페어와 상의하고 대체해보았어요. 
- 1주차 학습의 핵심인 테스트 코드를 잘 적용하기 위해 노력했습니다.
- 특히 `@ParameterizedTest` 혹은 `@RepeatedTest` 등을 이용하여 빈틈이 많이 없는 테스트를 만들기 위해 노력해보았어요.
		
저희는 위의 적혀있는 설계 포인트를 중요하게 생각하고 미션 진행해보았는데 아직까지 미숙한 부분이 많이 있을 것 같습니다.

잘 부탁드립니다.

### CODE REVIEW

File: src/main/java/domain/Car.java [hunk 1]
```diff
@@ -0,0 +1,47 @@
+package domain;
+
+public class Car {
+
+    private final String carName;
+    private Integer position;
+    private final PowerGenerator powerGenerator;
```
Comment (her0807) [inline src/main/java/domain/Car.java]: Car 객체를 잘 만들어주셨네요 👍🏻

요기서 쪼금 더 고민해보자면,` 자동차는 자동차의 이름을 검증하는 책임을 가지고 있어야할까요?`
자동차는 이동하는 것이 가장 큰 책임인 것 같은데` 중요한 로직이 검증 로직들에  가려져서 잘 안보이는 것 같아요. `
그래서 조금 더 원시값 포장을 해서 각 객체들이 각자의 할 일에 집중할 수 있게 해주면 어떨까요? 🙂

+ position 에 음수가 들어올 수 있을까요?

File: src/main/java/domain/Car.java [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+    public String getCarName() {
+        return this.carName;
+    }
+
+    public void tryMove() {
+        if (powerGenerator.generate().isSufficientPower()) {
+            move();
+        }
+    }
```
Comment (her0807) [inline src/main/java/domain/Car.java]: 🖐🏻 브루마블 해보셨나요? 
판에 말들이 있고, 주사위와 황금카드 등등을 확인하고 판에 말을 움직일지 말지 결정하는데요~

**자동차 경주에서도 움직일지 말지를 결정하는 주체가 필요할 것 같아요.** 

Power에서 랜덤 값이 4이상인지 검증하는 로직이 있는 것 같은데요. 
Power 라는 이름을 PowerStrategy 라는 것으로 변경하고 PowerGenerator 를 의존해서 
 그 객체에게 책임을 맡겨볼까요?

```java

    public void tryMove() {
        if (canMove()) {
            move();
        }
    }
```
그러면 내부 로직이 잘 가려지고(캡슐화) 책임이 적절하게 분리될 것 같아요.

File: src/main/java/domain/Car.java [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+    private void move() {
+        position++;
+    }
+
+    private void validate(String carName) {
+        validateCarNameLength(carName);
+    }
+
+    private void validateCarNameLength(String carName) {
+        if (carName.length() > 5) {
```
Comment (her0807) [inline src/main/java/domain/Car.java]: 클린코드에 `의미가 불분명한 매직 넘버를 상수로 선언하라.` 라는 내용이 있습니다! 🖐🏻

해당 숫자가 어떤 것을 의미하는지 적어두면 이후에 test 코드에서 재활용할 수도 있고, 다른 사람에게 정확한 의미 전달이 될 것 같아요오 ㅎㅎㅎ

File: src/main/java/domain/Car.java [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+        }
+    }
+
+    private void move() {
+        position++;
+    }
+
+    private void validate(String carName) {
+        validateCarNameLength(carName);
+    }
```
Comment (her0807) [inline src/main/java/domain/Car.java]: 검증 로직을 메서드로 추출한 것 💯

때로는 불필요한 메서드 분리가 가독성을 떨어트리고 로직을 복잡하게 할 때가 있는것 같아요 ~ 
저는 행위가 여러개일 경우 메서드로 묶곤 합니다! 요기처럼 validate 내부에 검증이 하나일 때는 
메서드로 감싸지 않아도 될 것 같아요!

File: src/test/java/domain/MoveCountTest.java [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+                .isInstanceOf(IllegalArgumentException.class);
+    }
+
+    @DisplayName("자동차 이동 횟수가 양의 정수일 경우 도메인이 생성된다.")
+    @ParameterizedTest
+    @ValueSource(ints = {1, 3, 5})
+    void testCreateMoveCountWithValidValue(int moveCount) {
+        assertThatCode(() -> MoveCount.from(moveCount)).doesNotThrowAnyException();
+    }
+
```
Comment (her0807) [inline src/test/java/domain/MoveCountTest.java]: 단위 테스트를 잘 작성해주셨네요 👍🏻

제가 예전에 처음 테스트를 작성했을 때에 비하면 엄청 잘 작성하신 것 같아요오 ㅎㅎ
이렇게 성공하거나 실패하는 테스트를 작성할 때 `어떤 값으로 검증해야하지?` 가 고민 됬던 것 같은데 
경계값을 사용해서 테스트 한다면, 불필요한 테스트를 줄이면서 신뢰도 높은 테스트를 작성할 수 있어요!

### 참고자료
[경계값 분석 테스트
(boundary value analysis test)](https://staedtler1207.tistory.com/21#google_vignette)

File: src/test/java/util/StringParserTest.java [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+@DisplayName("문자열 가공 기능 테스트")
+public class StringParserTest {
+
+    @DisplayName("지정해준 구분자를 바탕으로 문자열을 쪼갤 수 있다")
+    @Test
+    void testSplit() {
+        String target = "리비,에버,테니,잉크,제리";
+        List<String> split = StringParser.split(target, ",");
+        assertThat(split).containsExactly("리비", "에버", "테니", "잉크", "제리");
+    }
```
Comment (her0807) [inline src/test/java/util/StringParserTest.java:21]: 아래처럼 공백이 들어가면 어떻게 될까요? 🤔

```java

 String target = "리비.   ,에버,      테니,잉크,제리";
```

File: src/main/java/controller/RacingGameController.java [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+
+import domain.Car;
+import domain.Cars;
+import domain.MoveCount;
+import java.util.List;
+import util.StringParser;
+import view.InputView;
+import view.OutputView;
+
+public class RacingGameController {
```
Comment (her0807) [inline src/main/java/controller/RacingGameController.java]: ### 에버! controller 는 어떤 용도로 만들어졌나요? 🔥

MVC 패턴에 Controller 라면 사용자 입력을 처리하고, 애플리케이션의 흐름을 관리하는 용도로 쓰이면 좋을 것 같은데요 ~

지금은 자동차들을 입력받아서 경주를 하기 위해 준비시키고, 경주를 진행하고,  결과 값을 출력하는 등 많은 역할을 가지고 있는 것 같아요.  

MVC 패턴을 사용하는 이유도 결국 `1) 카운터 : 요청을 받고 응답을 하는 역할`, `2) 요리사:  비지니스를 담당하는 역할`, `3)만들어진 음식 : 사용자에 보여지는 결과물 ` 각 구성 요소의 역할과 책임을 명확하게 구분하기 위해서라고 생각합니다. 

그래서 도메인에게 적절한 책임이 무엇인가를 많이 고민해봤으면 좋겠어요! 💓

Comment (her0807) [reply src/main/java/controller/RacingGameController.java]: 지금은 console 이고 webView 나 app 이 없기 때문에 
근본적으로 controller 가 필요한지도 고민해보면 좋을것 같고요!

File: src/main/java/controller/RacingGameController.java [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+
+    private MoveCount prepareMoveCount() {
+        try {
+            Integer count = StringParser.parseToInt(inputView.requestMoveCount());
+            return MoveCount.from(count);
+        } catch (IllegalArgumentException e) {
+            return prepareMoveCount();
+        }
+    }
+}
```
Comment (her0807) [inline src/main/java/controller/RacingGameController.java]: -1 음수를 넣었는데, 바로 다시 시작하더라고요. 예외에는 적절한 메세지가 있으면 좋을 것 같아요!

**Review by her0807 — CHANGES_REQUESTED** [review]
안녕하세요, 에버 🫡

리뷰어 수달입니다!

남겨주신 코멘트 중에 `다만 원칙을 지킴으로써 얻는 코드 퀄리티보다 코드 복잡성이 크다고 판단한 경우는 페어와 상의하고 대체해보았어요.` 라는 부분 정말 인상 깊네요 👍🏻

코드는 항상 `trade-off` 가 존재하는 것 같아요.  어떤 점에서 원칙을 지켰을 때 복잡성이 올라간다고 생각하셨나요? 같이 얘기해보면 재밌을것 같아요오 -! 😄 

코드 보고 간단한 질문 남겼습니다! 재밌게 리뷰 나눠요 🙇🏻‍♀️

File: src/main/java/domain/Car.java (hunk 4 above)
Comment (helenason) [reply src/main/java/domain/Car.java]: 이후 추가될 가능성이 있는 검증 로직을 고려하여 선택했던 방법이었는데,
지금 생각해보니 너무 멀리 생각한 것 같네요.
다음 미션부터는 미션 내용에 조금 더 집중해보겠습니다! 😁

**Review by her0807 — APPROVED** [review]
어제 늦은시간까지 작업해주셨군요 에버 :)

리뷰가 잘 반영된 것 같아 이만 머지하겠습니다 수고 많으셨어요!
추가로 2단계에서는 에버의 생각을 코멘트로 더 많이 남겨주시면 좋을 것 같아요.
